📝 Add initial sphinx ext for per-doc feedbacks

webknjaz opened the pull request:
Resolves #8783

pradyunsg commented:
https://pip--8848.org.readthedocs.build/en/8848/

Uhhh... This is extremely invasive IMO.

webknjaz commented:
@pradyunsg I thought so. But atm I'm just doing what I was asked to do. It's easier to reason about stuff if there's preview :)

webknjaz commented:
I've colored it a bit. I'll let @nlhkabu / @jdittrich / @georgiamoon decide how to deal with the invasiveness.

FWIW it's a good start and can be tweaked.

webknjaz commented on news/8848.doc:
```diff
@@ -0,0 +1 @@
+8783.doc
```
N.B. It's a symlink to the other fragment.

pradyunsg replied:
Would it be worthwhile to we have the same text copied over instead? That might just be less fragile if someone's on Windows or whatever. :3

webknjaz replied:
Does anybody on windows generate changelogs? I'd address this only if such a problem actually occurs...

pfmoore replied:
Yes, I would do so if I were the one making a release. IMO, we shouldn't have symlinks in the pip repository. If that's not already a project standard, I'd be inclined to make it one. Maintainers do their best to make sure the release process works on Windows, but there's always the odd glitch that I need to fix - so anything we can do to reduce the risk of such things happening, we should IMO.

Is there a pre-commit hook we can add to check for and reject symlinks in the repo?

webknjaz replied:
I only saw a hook that checks that symlinks point to things that actually exist...

P.S. Just out of curiosity, could you please check if the symlink actually breaks anything on Windows? (I don't think it does but still, it'd be good to know)

webknjaz replied:
Oh, and the reason I'm using a symlink is just that if somebody wants to update the message, it'd be updated in both places, not just one.

pfmoore replied:
> Just out of curiosity, could you please check if the symlink actually breaks anything on Windows? 

Not really right at the moment, I'm not set up to try running the release scripts right now. I'll see if I can set something up at some point.

BTW, how does git even do symlinks if you check out on a filesystem that doesn't support them?

webknjaz replied:
Looks like it just turns them into text files
https://stackoverflow.com/questions/5917249/git-symlinks-in-windows

jdittrich commented:
> I'll let @nlhkabu / @jdittrich / @georgiamoon decide how to deal with the invasiveness.

Imho it is too invasive on such a short page, but totally OK on a longer page. We could hide one in case of short pages (don't know how hard this is). We could also make the box less big in general: 

> ![image](https://user-images.githubusercontent.com/3416487/92326422-50b14200-f052-11ea-9783-84c685732af8.png)

McSinyx commented on docs/docs_feedback_sphinxext.py:
```diff
@@ -0,0 +1,97 @@
+        )
+
+    questions_list = """
+      1. What problem were you trying to solve when you came to this page?
+
+      2. What content was useful?
+
+      3. What content was not useful?
```
```suggestion
    1. What problem were you trying to solve when you came to this page?
    2. What content was useful?
    3. What content was not useful?
```

The rendered list seems to be over-indented, probably because triple-quotes preserve whitespace.  I can't view the reST output (after this extension is applied) due to the theme, but I think generally it'd be safer to declare all these literals as module level constants to avoid accidentally messing up indentation.

webknjaz replied:
I hear you but because it's also augmented with another line with context-specific var, I'm not sure if moving it on the module level is a good idea. After all, we need to keep some sort of common indentation level here and 3 lines below.

I guess I'll just experiment with dedenting both.

pradyunsg replied:
I'd say: check if it's creating a blockquote in the final HTML. If it is, then you know that things are indented too far.

McSinyx replied:
It is, and I was also concern about other texts provided as indented triple quoted literals, e.g. **Did this article help?** (relative to `.. important`).  Sphinx is pretty relaxed and still renders the content even if it doesn't align with `important`, but I *think* it'd be easier to notice whitespace issues if the reST is provided as something close to unindented literals, which we can then str.format on.

pfmoore commented:
I agree with @pradyunsg - this is too intrusive. Top or bottom is fine, but I'm not keen on both.

(Also, I think having the banner on the changelog page is probably a bit much, although I understand that it's an all-or-nothing thing because of the way it's implemented).

pradyunsg commented:
It's not an all-or-nothing proposition since we have the document name in the code that's adding the $thing. We can have conditionals to "opt out" for certain pages, if we want.

webknjaz commented:
> It's not an all-or-nothing proposition since we have the document name in the code that's adding the $thing. We can have conditionals to "opt out" for certain pages, if we want.

Exactly, we can change this once there's a general consensus on what exactly is needed.

webknjaz commented:
> We could hide one in case of short pages (don't know how hard this is).

The easiest way of doing this is to maintain a list of exclusions I think.

> We could also make the box less big in general

Feel free to send a CSS patch with a specific suggestion. "less big" is too broad and is not very well actionable...

webknjaz commented:
@jdittrich oh, I just looked at the picture again. Are you suggesting to drop the list?

jdittrich commented:
> @jdittrich oh, I just looked at the picture again. Are you suggesting to drop the list?

Yes, that would be one way to make it smaller (which could be particularly important for the top box)

webknjaz commented:
Alright, so just to clarify the action items now:
1) drop the list from the top box, but not the bottom one
2) dedent the list
3) implement page exclusions (not sure if everybody agrees on this one)

Anything I'm missing?

webknjaz commented:
So I think I've made it a bit better:

1. Now the list is only present in the bottom box but not the top one (as requested by @jdittrich)
2. It is possible to exclude documents and I've added the `news` document to the exclude list (:wave: @pfmoore)
3. Adjusted the indentation (@McSinyx :v:)
4. The top box is hidden if the document RST source has over 50 LOC (@jdittrich :tada:)
5. I've added a number of external toggles so that it's possible to adjust the following w/o the need to look into the monstrous extension implementation:
  * box (admonition) type
  * the "big doc" threshold
  * the target email
  * the excluded documents set
  * the questions list

pradyunsg commented:
I like it. @nlhkabu and @jdittrich if you reckon this is good to go, let's ship it. :)

nlhkabu commented:
Looks good to me! Thanks @webknjaz ! :heart:

jdittrich commented:
thanks!
fine for me, too :)
